server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -483,6 +489,76 @@ public DesiredBalance compute(
         return new DesiredBalance(lastConvergedIndex, assignments, routingNodes.getBalanceWeightStatsPerNode(), finishReason);
     }
 
+    /**
+     * For shards started after initial polling of the ClusterInfo but before the next polling, we need to
+     * account for their impacts by simulating the events, either relocation or new shard start. This is done
+     * by comparing the current RoutingNodes against the shard allocation information from the ClusterInfo to
+     * find out the shard allocation changes. Note this approach is approximate in some edge cases:
+     * <ol>
+     * <li> If a shard is relocated twice from node A to B to C. It is considered as relocating from A to C directly
+     * for simulation purpose.</li>
+     * <li>If a shard has 2 replicas and they both relocate, replica 1 from A to X and replica 2 from B to Y. The
+     * simulation may see them as relocations A->X and B->Y. But it may also see them as A->Y and B->X. </li>
+     * </ol>
+     * In both cases, it should not really matter for simulation to account for resource changes.
+     */
+    static void maybeSimulateAlreadyStartedShards(
+        ClusterInfo clusterInfo,
+        RoutingNodes routingNodes,
+        ClusterInfoSimulator clusterInfoSimulator
+    ) {
+        // Find all shards that are started in RoutingNodes but have no data on corresponding node in ClusterInfo
+        final var startedShards = new ArrayList<ShardRouting>();
+        for (var routingNode : routingNodes) {
+            for (var shardRouting : routingNode.started()) {
+                // We use dataPath to find out whether a shard is allocated on a node.
+                // TODO: DataPath is sent with disk usages but thread pool usage is sent separately so that local shard allocation
+                // may change between the two calls.
+                if (clusterInfo.getDataPath(shardRouting) == null) {
```

Comment:
NIt: I wonder if we could put this logic in the `ClusterInfo` and have the computer be more agnostic about how this is implemented.

i.e. if we implemented `boolean ClusterInfo#shardHasMoved(ShardRouting)` and then put this logic and an explanation of it in there. It might reduce the cognitive load for the reader of this method, and allow us to adapt the implementation of `#shardHasMoved` as the contents of `ClusterInfo` evolves.

Reply:
Good suggestion. Pushed 7bad062afcbb10e0c5aef52f05d210b27c06bd4b Thanks!